### PR TITLE
Toast docs + children support

### LINKED ISSUE #1224: Toast api seems to violate core React principles
### Discussion
- __Package version(s): 1.17.1

[Toasts and the Toaster component](http://blueprintjs.com/docs/#core/components/toast) api do not to follow standard React data flow. Most of the blueprint component api (and indeed most React components) allow data and configuration to be passed to components via props. The Toaster component api relies on the methods exposed on a reference to the component in order to manage the component. As evidenced by #1205, the dom manipulation underneath `Toaster.create` and `Toaster.show` are also not intuitive React behavior - they manipulate the dom outside of the react render cycle.

My guess is that the behavior of self-dismissing, ephemeral toasts is part of the design of the component, and that behavior doesn't sync well with a component api based exclusively on props.

### Toaster api from docs
```
import { Button, Position, Toaster } from "@blueprintjs/core";
 
class MyComponent extends React.Component<{}, {}> {
    private toaster: Toaster;
    private refHandlers = {
        toaster: (ref: Toaster) => this.toaster = ref,
    };
 
    public render() {
        return (
            <div>
                <Button onClick={this.addToast} text="Procure toast" />
                <Toaster position={Position.TOP_RIGHT} ref={this.refHandlers.toaster} />
            </div>
        )
    }
 
    private addToast = () => {
        this.toaster.show({ message: "Toasted!" });
    }
}
```

### What I would expect from a 'standard' react component api
```
import { Button, Position, Toaster } from "@blueprintjs/core";
 
class MyComponent extends React.Component<{}, {}> { 
    constructor(props) {
        super(props)
        this.state = {
            toasts: []
        }
   }

    public render() {
        return (
            <div>
                <Button onClick={this.addToast} text="Procure toast" />
                <Toaster position={Position.TOP_RIGHT} toasts={this.state.toasts} />
            </div>
        )
    }
 
    private addToast = () => {
        this.setState({
             toasts: [{message: "Toasted!" }]
        });
    }
}
```

I'd love to know more about the design decisions that went into this component, how it is used or intended to be used in apps where toasts are stored in data, and what the future of this component looks like. Will there be a Toaster component exposed that takes messages as props? I'd like to manage toast data independent of how it is rendered.

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

- `Toaster` renders `children` after state toasts
- 🐛  `Overlay` now passes through non-JSX `children`
  - `string | number` children will not be animated. to be fair, this is an improvement over the previous behavior where it would simply die if given non-JSX children.
- major refactors to Toasts documentation to clarify expected usage (see #1224)

### CODE REVIEW

**Comment by giladgray** [comment]
#### Question:
Should `Toaster` support a `toasts` _prop_ just like its state? 
And I'll use lifecycle methods to keep them in sync. 
Or is `children` enough? 

(my gut is `children` is enough.)

File: packages/core/src/components/toast/toast.md [hunk 1]
```diff
@@ -58,51 +51,71 @@ OurToaster.update(key, { message: "Still toasted!" });
...
 
-@### React component
+@## React component usage
 
-The `Toaster` React component is a stateful container for a single list of toasts. Internally, it
-uses [`Overlay`](#core/components/overlay) to manage children and transitions. It can be vertically
-aligned along the top or bottom edge of its container (new toasts will slide in from that edge) and
-horizontally aligned along the left edge, center, or right edge of its container.
+Render the `<Toaster>` component like any other element. Then, either access the instance methods by supplying
+a `ref` handler (see example below), or supply rendered `<Toast>` elements as `children` of the `<Toaster>`.
```
Comment (adidahiya) [inline packages/core/src/components/toast/toast.md]: I'm not sure why we're suggesting a `ref` handler... if someone really wanted to do this in a React-idiomatic way, wouldn't they provide toasts as `children`? What are the benefits of this over the `Toaster.create()` API? I feel all these options make the usage docs more confusing.

File: packages/core/src/components/toast/toast.md [hunk 2]
```diff
@@ -58,51 +51,71 @@ OurToaster.update(key, { message: "Still toasted!" });
     enable `autoFocus` for a `Toaster` via a prop, if desired.
 </div>
 
-@### Static method
+@interface IToaster
```
Comment (adidahiya) [inline packages/core/src/components/toast/toast.md]: I think this should go inside the "Static usage" section, after the first bit of prose there. Prefer reading prose first, detailed props tables later

File: packages/core/src/components/toast/toast.md (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/toast/toast.md]: i'll move the `children` suggestion first and note that you can `ref` but please don't (see next paragraph).

File: packages/core/src/components/toast/toast.md [hunk 3]
```diff
@@ -20,29 +13,29 @@ have more than one toast onscreen at a time.
...
+
+The `Toaster` React component is a stateful container for a single list of toasts. Internally, it
+uses [`Overlay`](#core/components/overlay) to manage children and transitions. It can be vertically
+aligned along the top or bottom edge of its container (new toasts will slide in from that edge) and
+horizontally aligned along the left edge, center, or right edge of its container.
+
+`Toaster` can be used in one of two ways:
+
+1. `Toaster.create(props)` static method returns a new `IToaster` instance. Use the instance method `toaster.show()` to manipulate this instance. __(recommended)__
+1. Render a `<Toaster>` element and either use the `ref` prop to access its instance methods, or pass rendered `<Toast>`s as `children`.
```
Comment (adidahiya) [inline packages/core/src/components/toast/toast.md]: This feels like 3 options.

1. `Toaster.create(...)`
2. `<Toaster ref={...} />`
3. `<Toaster><Toast ... /></Toaster>` 

they can all be mixed & matched anyway, so might as well list all 3.

Comment (giladgray) [reply packages/core/src/components/toast/toast.md]: 2 and 3 can be mixed, no way to `provide` children to 1.

but yes there are three, so i'll update

Comment (adidahiya) [reply packages/core/src/components/toast/toast.md]: well you can combine option 1 with 2 & 3, you'd just get multiple Toaster instances

**Comment by rrcobb** [comment]
Love it, thanks for following through!

**Pull request merged**